omniduct/caches/base.py

```diff
@@ -99,26 +96,115 @@ def __init__(self, **kwargs):
     def _init(self):
         pass
 
-    @abstractmethod
-    def clear(self, id_duct, id_str):
+    # Data insertion and retrieval
+
+    def set(self, key, value, namespace=None, serializer=PickleSerializer, expires=None, metadata=None):
+        namespace = self._namespace(namespace)
+        key = self._key(key)
+        # try:
+        self.set_metadata(key, metadata, namespace=namespace, replace=True)
+        with self._get_stream_for_key(namespace, key, 'data{}'.format(serializer.file_extension), mode='wb', create=True) as fh:
+            return serializer.serialize(value, fh)
+        # except:
+        #     self.unset(key, namespace=namespace)
+
+    def set_metadata(self, key, metadata, namespace=None, replace=False):
+        namespace = self._namespace(namespace)
+        key = self._key(key)
+        if replace:
+            orig_metadata = {'created': datetime.datetime.utcnow()}
+        else:
+            orig_metadata = self.get_metadata(key, namespace=namespace)
+
+        orig_metadata.update(metadata or {})
+
+        with self._get_stream_for_key(namespace, key, 'metadata', mode='w', create=True) as fh:
+            yaml.safe_dump(orig_metadata, fh, default_flow_style=False)
+
+    def get(self, key, namespace=None, serializer=PickleSerializer):
+        namespace = self._namespace(namespace)
+        key = self._key(key)
+        try:
+            with self._get_stream_for_key(namespace, key, 'data{}'.format(serializer.file_extension), mode='rb', create=False) as fh:
+                return serializer.deserialize(fh)
+        finally:
+            self.set_metadata(key, namespace=namespace, metadata={'last_accessed': datetime.datetime.utcnow()})
+
+    def get_metadata(self, key, namespace=None):
+        namespace = self._namespace(namespace)
```

Comment:
Sorry... not following >.<. Are you suggesting:
`namespace, key = self._namespace(namespace), self._key(key)`?

Reply:
I'd go a step further and say since we're not reusing these variables any where, we should just use them inline 
```
self._get_stream_for_key(self._namespace(namespace), self._key(key), 'metadata', mode='r', create=True)
...
```